code/WorkInProgress/racing.dm

```diff
@@ -191,158 +191,145 @@
 			R.random_powerup()
 			qdel(src)
 
-/obj/powerup
-	name = "powerup"
-	desc = "Click to use"
+/// This holds all of your abilities for kart racing
+/datum/abilityHolder/kart_racing
+	topBarRendered = 1
+
+/// the base kart ability
+ABSTRACT_TYPE(/datum/targetable/kart_powerup)
+/datum/targetable/kart_powerup
 	icon = 'icons/misc/racing.dmi'
 	icon_state = "blank"
-	anchored = ANCHORED
-	layer = HUD_LAYER
-	screen_loc = "NORTH,WEST"
-	var/obj/racing_clowncar/owner
-
-	disposing()
-		if(owner?.powerup == src)
-			if(owner?.driver?.client)
-				owner.driver.client.screen -= src
-			owner.powerup = null
-		owner = null
-		..()
-
-	Click()
-		if (owner.powerup == src)
-			owner.powerup = null
-		qdel(src)
+	desc = "Click to use"
+	cooldown = 0
+	last_cast = 0
+	targeted = 0
+	preferred_holder_type = /datum/abilityHolder/kart_racing
+	var/mob/living/carbon/human/racer
+
+	onAttach(datum/abilityHolder/H)
+		. = ..()
+		if(ishuman(holder.owner))
+			racer = holder.owner
 		return
 
-/obj/powerup/bananapeel
+	cast()
+		if (!istype(racer))
+			return 1
+		if (..())
+			return 1
+		if(!istype(racer?.loc,/obj/racing_clowncar))
+			return 1
+
+/datum/targetable/kart_powerup/banana_peel
 	name = "Bananapeel"
 	desc = "Click to use"
-	anchored = ANCHORED
 	icon_state = "banana"
-	screen_loc = "NORTH,WEST"
-
-	Click()
 
-		if(!istype(src.loc,/obj/racing_clowncar))
-			qdel(src)
-			return
+	cast()
+		if (..())
+			return 1
 
-		var/turf/T = get_turf(src.loc)
+		var/turf/T = get_turf(get_turf(racer))
 		new/obj/racing_trap_banana/(T)
 
 		playsound(T, 'sound/mksounds/itemdrop.ogg', 45, FALSE)
-
+		// remove ourselves when we're used
+		racer.removeAbility(src)
+		holder.abilities -= src
+		racer = null
 		qdel(src)
```

Comment:
```suggestion
		racer.removeAbilityInstance(src)
```
This is the correct proc, the `qdel` and `abilities -=` are handled by the ability holder and the `racer = null` should be in the parent `disposing`. Same goes for all other instances of this snippet.